[SQLair] Replace slice queries in domain/cloud/state

Aflynn50 opened the pull request:
<!-- Why this change is needed and what it does. -->
Use SQLair for slice queries in domain/cloud/state. Other queries in the same transactions are updated to use SQLair however there are some that could benefit from planned SQLair features and may benefit from being updated once those are added.
## Checklist

<!-- If an item is not applicable, use `~strikethrough~`. -->

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
~- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
~- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps
```TEST_PACKAGES="./domain/cloud/state -count=1 -race" TEST_FILTER="Suite" make run-go-tests```
<!-- Describe steps to verify that the change works. -->

## Documentation changes
N/A
<!-- How it affects user workflow (CLI or API). -->

SimonRichardson commented:
/build

wallyworld commented on domain/cloud/state/state.go:
```diff
@@ -640,97 +640,107 @@ func updateAuthTypes(ctx context.Context, tx *sql.Tx, cloudUUID string, authType
+
 	for _, a := range authTypeIds {
-		if _, err := tx.ExecContext(ctx, insertQuery, utils.MustNewUUID().String(), cloudUUID, a); err != nil {
+		if err := tx.Query(ctx, insertStmt, CloudAuthType{ID: utils.MustNewUUID().String(), CloudUUID: cloudUUID, AuthTypeID: a}).Run(); err != nil {
```
It would be less cluttered/messy to make the CloudAuthType struct first and then pass that to the query instead of creating the struct inline.

wallyworld commented on domain/cloud/state/state.go:
```diff
@@ -640,97 +640,107 @@ func updateAuthTypes(ctx context.Context, tx *sql.Tx, cloudUUID string, authType
+	insertQuery, err := sqlair.Prepare(`
 INSERT INTO cloud_region (uuid, cloud_uuid, name, endpoint, identity_endpoint, storage_endpoint)
-VALUES (?, ?, ?, ?, ?, ?)
+VALUES ($CloudRegion.uuid, $CloudRegion.cloud_uuid, $CloudRegion.name, $CloudRegion.endpoint, $CloudRegion.identity_endpoint, $CloudRegion.storage_endpoint)
```
maybe split this long line

wallyworld commented on domain/cloud/state/state.go:
```diff
@@ -640,97 +640,107 @@ func updateAuthTypes(ctx context.Context, tx *sql.Tx, cloudUUID string, authType
 	for _, cert := range certs {
-
-		if _, err := tx.ExecContext(ctx, insertQuery, utils.MustNewUUID().String(), cloudUUID, cert); err != nil {
+		if err := tx.Query(ctx, insertQuery, CloudCACert{ID: utils.MustNewUUID().String(), CloudUUID: cloudUUID, CACert: cert}).Run(); err != nil {
```
Same here: it would be less cluttered/messy to make the CloudCACert struct first and then pass that to the query instead of creating the struct inline.

Aflynn50 replied:
Refactored, also refactored other places where this was the case.

wallyworld reviewed (approved):
Thanks for fixing this. Looks ok to land after once pre landing tests pass.

Aflynn50 commented:
PR updated to include changes to `loadClouds` function. I had previously left it unchanged as I didn't want to change the SQL and SQLair would not work with the query in its current form. After speaking with @manadart I've refactored the SQL and updated the transaction to use SQLair.

SimonRichardson commented on domain/cloud/state/state.go:
```diff
@@ -332,62 +334,65 @@ ON CONFLICT(region_uuid, key) DO UPDATE
 			cld = &cloud.Cloud{
 				Name:              dbCloud.Name,
-				Type:              cloudType,
+				Type:              m["cloud_type"].(string),
```
This can potentially panic. Check it exists before using it.

Aflynn50 replied:
Added an `ok` check. I was unsure about this when writing it, technically SQLair is guaranteeing that this value is there, it would have returned an error when the query was executed if it was not, but I agree we should avoid any potential for panics at all here.

wallyworld commented on domain/cloud/state/types.go:
```diff
@@ -74,10 +93,36 @@ type CloudRegion struct {
+	Value string `db:"value"`
+}
+
+type CloudUUIDs []string
```
Don't agree we need to comment all these indivually. The types file has a global comment at the top

`// These structs represent the persistent cloud entity schema in the database.`

Their use is quite obvious

SimonRichardson commented on domain/cloud/bootstrap/bootstrap.go:
```diff
@@ -20,7 +21,7 @@ import (
 	return func(ctx context.Context, db database.TxnRunner) error {
-		return errors.Trace(db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
 			cloudUUID, err := uuid.NewUUID()
```
I know you didn't do this, but whilst you're here, can you move the `cloudUUID, err := uuid.NewUUID()` line outside of the `Txn` method.

Aflynn50 replied:
Changed

SimonRichardson commented on domain/cloud/state/state.go:
```diff
@@ -396,59 +397,66 @@ ON CONFLICT(region_uuid, key) DO UPDATE
 	if name != "" {
-		q = q + "WHERE cloud.name = ?"
-		args = append(args, name)
+		q = q + "WHERE cloud.name = $M.cloud_name"
```
```suggestion
		q += "WHERE cloud.name = $M.cloud_name"
```

SimonRichardson reviewed (commented):
Last tweak.

Aflynn50 commented:
/merge
